clang/runtime/dpct-rt/include/sparse_utils.hpp

```diff
@@ -743,6 +766,311 @@ inline void spmm(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                             "attributes to be specified.");
+  }
+  oneapi::mkl::sparse::optimize_trsv(
+      queue, a->get_uplo().value(), oneapi::mkl::transpose::nontrans,
+      a->get_diag().value(), a->get_matrix_handle());
+}
+
+/// Solves a system of linear equations for a sparse matrix.
+/// \param [in] queue The queue where the routine should be executed. It must
+/// have the in_order property when using the USM mode.
+/// \param [in] trans_a Specifies operation on input matrix a.
+/// \param [in] alpha Specifies the scalar alpha.
+/// \param [in] a Specifies the sparse matrix a.
+/// \param [in] x Specifies the dense vector x.
+/// \param [in, out] y Specifies the dense vector y.
+/// \param [in] data_type Specifies the data type of \param a, \param x and
+/// \param y .
+inline void spsv(sycl::queue queue, oneapi::mkl::transpose trans_a,
+                 const void *alpha, sparse_matrix_desc_t a,
+                 std::shared_ptr<dense_vector_desc> x,
+                 std::shared_ptr<dense_vector_desc> y,
+                 library_data_t data_type) {
+  switch (data_type) {
+  case library_data_t::real_float:
+    detail::check_alpha_value<float>(alpha, queue);
+    break;
+  case library_data_t::real_double:
+    detail::check_alpha_value<double>(alpha, queue);
+    break;
+  case library_data_t::complex_float:
+    detail::check_alpha_value<std::complex<float>>(alpha, queue);
+    break;
+  case library_data_t::complex_double:
+    detail::check_alpha_value<std::complex<double>>(alpha, queue);
+    break;
+  default:
+    throw std::runtime_error("Unsupported data type.");
+  }
+
+  if (!a->get_uplo() || !a->get_diag()) {
+    throw std::runtime_error("oneapi::mkl::sparse::trsv needs uplo and diag "
+                             "attributes to be specified.");
+  }
+  oneapi::mkl::uplo uplo = a->get_uplo().value();
+  oneapi::mkl::diag diag = a->get_diag().value();
+
+  std::uint64_t key = dpct::detail::get_type_combination_id(
+      a->get_value_type(), x->get_value_type(), y->get_value_type());
+  switch (key) {
+  case dpct::detail::get_type_combination_id(library_data_t::real_float,
+                                             library_data_t::real_float,
+                                             library_data_t::real_float): {
+    auto data_x =
+        dpct::detail::get_memory(reinterpret_cast<float *>(x->get_value()));
+    auto data_y =
+        dpct::detail::get_memory(reinterpret_cast<float *>(y->get_value()));
+    oneapi::mkl::sparse::trsv(queue, uplo, trans_a, diag,
+                              a->get_matrix_handle(), data_x, data_y);
+    break;
+  }
+  case dpct::detail::get_type_combination_id(library_data_t::real_double,
+                                             library_data_t::real_double,
+                                             library_data_t::real_double): {
+    auto data_x =
+        dpct::detail::get_memory(reinterpret_cast<double *>(x->get_value()));
```

Comment:
add a template version of get_memory accept void * and do reinterpret_cast in it.

Reply:
done